scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala

```diff
@@ -279,185 +240,226 @@ class ScalametaParser(input: Input)(implicit dialect: Dialect) { parser =>
     case Tokenized.Success(tokens) => tokens
     case Tokenized.Error(_, _, details) => throw details
   }
-  lazy val (parserTokens, parserTokenPositions) = {
-    def indentedRegion(regions: List[SepRegion]): Boolean =
-      regions.headOption.exists(_.isIndented)
 
-    val parserTokens = mutable.ArrayBuilder.make[Token]
-    val parserTokenPositions = mutable.ArrayBuilder.make[Int]
+  // NOTE: public methods of TokenIterator return scannerTokens-based positions
+  trait TokenIterator extends Iterator[Token] {
+    def prevTokenPos: Int
+    def tokenPos: Int
+    def token: Token
+    def fork: TokenIterator
+    def observeIndented(): Boolean
+    def observeIndentedEnum(): Boolean
+  }
+  var in: TokenIterator = {
+    new NewTokenIterator(
+      scannerTokens,
+      List.empty,
+      TokenRef(scannerTokens(0), 0, 1),
+      -1
+    )
+  }
+
+  case class TokenRef(
+      token: Token,
+      pos: Int,
+      nextPos: Int
+  )
+
+  class NewTokenIterator(
+      scannerTokens: Tokens,
+      var sepRegions: List[SepRegion],
+      var curr: TokenRef,
+      var prevPos: Int
+  ) extends TokenIterator {
+
+    override def hasNext: Boolean = curr.token.isNot[EOF]
+
+    override def next(): Token = {
+      val (newSepRegions, newTokenRef) = nextToken(curr.pos, curr.nextPos, sepRegions)
+      prevPos = curr.pos
+      curr = newTokenRef
+      sepRegions = newSepRegions
+      //println(s"return ${curr.token} ${curr.token.getClass()} $sepRegions ${hashCode()}")
+      curr.token
+    }
+
+    private def observeIndented0(f: (Int, List[SepRegion]) => List[SepRegion]): Boolean = {
+      if (!dialect.allowSignificantIndentation) false
+      else {
+        val existingIndent = sepRegions.find(_.isIndented).map(_.indent).getOrElse(0)
+        val hasLF = (prevPos to tokenPos).exists(i => scannerTokens(i).is[LF])
+        val expected = countIndent(tokenPos)
+        if (hasLF && expected > existingIndent) {
+          val prependIndentRegion = expected != existingIndent
+          if (prependIndentRegion) {
+            sepRegions = f(expected, sepRegions)
+            curr = TokenRef(
+              new Indentation.Indent(token.input, token.dialect, token.start, token.end),
+              curr.pos,
+              curr.pos
+            )
+          }
+          true
+        } else false
+      }
+    }
+
+    def observeIndented(): Boolean = {
+      observeIndented0 { (i, prev) =>
+        val undoRegionChange =
+          prev.headOption match {
+            case Some(RegionParen) if token.is[LeftParen] => prev.tail
```

Comment:
We should have `RegionIndent` and the `RegionParen` next - instead of just dropping it.

Reply:
It will be applied one more time.
After `observerIndented` were called the next token after `Indent` will be `(` that will put `RegionParen` again.

For example, in the code below at the moment when `observeIndented` is called iterator already reached `(` token on the second line and updated `sepRegions` on `RegionParen :: Nil`.

```scala
if (cond)
  <(>f)
```